Port `scatter` and `scatter_add` to `OpInfo`

ysiraichi opened the pull request:
Fixes #54302
Tracking Issue #54261 

**Summary:**
- Port `scatter` and `scatter_add` tests to `OpInfo`
- `masked_scatter` was already ported to `OpInfo`

ysiraichi committed:
Initial porting of scatter operations.

ysiraichi committed:
Split test for `scatter` only.

ysiraichi committed:
Refactor sample input generation.

ysiraichi committed:
Remove trailing spaces.

heitorschueroff commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -5000,15 +5038,6 @@ def method_tests():
-        ('scatter', (), (0, torch.tensor(0, dtype=torch.int64), 2.5), 'scalar_all_dim0', (), [0]),
-        ('scatter_add', (M, S), (0, gather_variable((S, S), 1, M), (S, S)), 'dim0', (), [0]),
-        ('scatter_add', (M, S), (1, gather_variable((M, S // 2), 0, S), (M, S // 2)), 'dim1', (), [0]),
-        ('scatter_add', (), (0, torch.tensor(0, dtype=torch.int64), ()), 'scalar_all_dim0', (), [0]),
```
This case is missing in the new OpInfo tests

ysiraichi replied:
I'm not sure I get which ones are missing. As far as I understand, all four of them are covered by `sample_inputs_scatter_ops`. Am I missing something?

- Line 5006: https://github.com/pytorch/pytorch/blob/dd64ee83ad6678aaeaa9368c64e2698eeef45628/torch/testing/_internal/common_methods_invocations.py#L2558-L2562
- Line 5007 and 5008: https://github.com/pytorch/pytorch/blob/dd64ee83ad6678aaeaa9368c64e2698eeef45628/torch/testing/_internal/common_methods_invocations.py#L2542-L2552
- Line 5009: https://github.com/pytorch/pytorch/blob/dd64ee83ad6678aaeaa9368c64e2698eeef45628/torch/testing/_internal/common_methods_invocations.py#L2554-L2556

heitorschueroff commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2535,6 +2535,34 @@ def sample_inputs_kron(op_info, device, dtype, requires_grad):
+        SampleInput(_make(()), args=(0, torch.tensor(0, dtype=torch.long, device=device), _make(()))),
+    )
+
+    if op_info.name == 'scatter':
```
scatter_add should also be tested for scalar input

ysiraichi replied:
I don't think `scatter_add` supports scalar inputs. Am I missing something?
https://github.com/pytorch/pytorch/blob/0a541e23e17a2b9555673c0877e15abcfe49a173/aten/src/ATen/native/native_functions.yaml#L4858-L4867

heitorschueroff reviewed (commented):
Just some small changes, otherwise PR looks good.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2535,6 +2535,34 @@ def sample_inputs_kron(op_info, device, dtype, requires_grad):
+    def _make(shape, dtype=dtype, low=None, high=None):
+        return make_tensor(shape, device, dtype, low=low, high=high, requires_grad=requires_grad)
+
+    test_cases = (
```
Hey @ysiraichi! Thanks for porting these operations to OpInfos.

This programmatic enumeration is neat, but it might be nicer to separate the sample input functions for these two operations and just hand-write most of the logic for now. In addition, what about testing the reduce parameter of scatter? 

So the code for scatter look like:

```
def sample_inputs_scatter(op_info, device, dtype, requires_grad, **kwargs):
  def _tensor(shape):
    return make_tensor(shape, device, dtype, requires_grad=requires_grad)
  def _gather(shape, index_dim, max_indices):
    return gather_variable(shape,  index_dim, max_indices, device=device)

  test_cases = (
    (_tensor((M, S)), 0, _gather((S, S), 1, M), _tensor((S, S))),
     ...
  )

  samples = []
  for test_case in test_cases:
    for reduce in (None, "add", "multiply"):
      sample = SampleInput(test_case[0], args=test_case[1:], kwargs={'reduce': reduce}
      samples.append(sample)

  return samples
```

scatter_add can reuse most of that function, except it doesn't have a reduce argument and (as you note in one of your comments) may not support all the inputs that scatter does.

What are you thoughts? Is an approach like this interesting? Would you suggest modifying it further?

ysiraichi replied:
Sure! I guess making them more explicit might be better, indeed.
There is a thing with testing the reduce parameter of scatter, though. If I'm not mistaken, only its inplace variant `scatter_` supports it. Should I create another `OpInfo` specially for `scatter_`?

mruberry replied:
Interesting! That's great attention to detail.

If only scatter_ supports the "reduce" parameter then I suggest not testing in this PR, actually, but filing an issue for this behavior ("scatter_ only supports 'reduce' kwarg on inplace variant") and the testing gap, and then adding a comment with a link to that issue above this sample input function. As for the sample input function itself, it would just remove the "reduce" enumeration, I think.

How does that sound? Then a future PR could look at extending the tests or extending scatter's kwarg support.

ysiraichi replied:
That makes sense.
I will do that. Thanks!

ysiraichi committed:
Address suggestions.

- Make test cases explicit
- Use one `SampleInput`s generator per operation
- Add link to issue with missing test

ysiraichi committed:
Return list comprehension.

ysiraichi committed:
Fix linting errors.

ysiraichi commented:
I believe the CI failure is unrelated.

mruberry reviewed (approved):
Awesome! Thanks @ysiraichi; and thank you for filing the related issue.

@heitorschueroff, would you like to take another look?

heitorschueroff reviewed (approved):
Changes LGTM! Thanks for porting these to OpInfo. I'll land this.
